### PR TITLE
Account for type parameters in missing function codefix

### LINKED ISSUE #49693: Add Missing Function Declaration: Does Not Correctly Declare Generic Function
# Bug Report

The "Quick Feature" allows us to generate a missing a function, but doesn't correctly declare that function when its parameters include a type parameter.

See also [unexpected behavior when type narrowing in this example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDAJgUzLKBPAPAFQD4AKAZzQBtgAuRfASkQG8AoRNxAek7rkQCc0AB35wUICGkRQAFlIBGIAOYAaRAEUQMCAGtEAMRgAPRIUQBBFCkQBbGKVIwwSxKEiwEidBAoBDfr4eYBwwwIjEOEJocGHkVIgAvEmIAERgIDbyaPwpjKyIBYX84ACiAG4wFADKqGglwMBo0GSUwPQA3BwAviwcglAg-EhxwJ09LNyu4NDwSMVg5ZU16PWN0OEjtPiIAGSI6ZnZeRM8BbKiAO77aFcl-KL8xCn600H7cFDINkIUaDYYUDQKAAdLlOpMekA).

### 🔎 Search Terms

- add missing function declaration
- generic type parameter

### 🕗 Version & Regression Information

I am unable to find a version of TypeScript where this was not a problem. 

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDAJgUzLKBPAPAFQD4AKAZzQBtgAuRfASkQG8AoRNxAek7rkQCc0AB35wUICGkRQAFlIBGIAOYAaRAEUQMCAGtEAMRgAPRIUQBBFCkQBbGKVIwwSxKEiwEidBAoBDfr4eYBz84ACiAG4wFADKqGhhwMBo0GSUwPQsIWhQIPxI5FQA3CwAvlks3K7g0PBIoWCR0XHoicmphTR0jKxVsqIA7ohgaENh-KL8xABE+jVBw3BQyDZCFGg2GFBoKAB00-QlVeVAA)

### 💻 Code

```ts
function identity<T>(self: T) {
 
  // To reproduce the bug, Quick Fix > Add missing function declaration
  runEvilSideEffect(self)

  return self;
}
```

### 🙁 Actual behavior

TypeScript generated this function, which is an invalid function declaration:

```ts
function runEvilSideEffect(self: T) {
    throw new Error("Function not implemented.");
}
```

### 🙂 Expected behavior

```ts
function runEvilSideEffect<T> (self: T) {
    throw new Error("Function not implemented.");
}
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
In addition to any explicit type parameters, also accounts for any type parameters in the types of arguments.

~I noticed both callers of `typeToAutoImportableTypeNode` use `checker.getBaseTypeOfLiteralType`, so I deduplicated & moved that inside the function.~ Edit: no longer true as of #50004. Reverted.

Fixes #49693

### CODE REVIEW

**Review by sandersn — CHANGES_REQUESTED** [review]
Initial comments, although I still need to read getArgumentTypesAndTypeParameters more closely

File: src/services/codefixes/helpers.ts [hunk 1]
```diff
@@ -362,6 +382,42 @@ namespace ts.codefix {
...
+                argumentTypeParameters.add(argumentTypeParameter);
+            }
+        }
+
+        return { argumentTypeNodes, argumentTypeParameters: arrayFrom(argumentTypeParameters.values()) };
+    }
+
+    function getFirstTypeParameterName(type: Type): string | undefined {
+        if (type.flags & TypeFlags.Intersection) {
+            for (const subType of (type as IntersectionType).types) {
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/services/codefixes/helpers.ts:408]: #50005

File: src/services/codefixes/helpers.ts [hunk 2]
```diff
@@ -349,7 +348,28 @@ namespace ts.codefix {
...
+                    CharacterCodes.T + i <= CharacterCodes.Z
+                        ? String.fromCharCode(CharacterCodes.T + i)
+                        : `T${i}`
+                );
+            }
+        }
+
+        return map(
+            arrayFrom(usedNames.values()),
+            (usedName) => factory.createTypeParameterDeclaration(/*modifiers*/ undefined, usedName),
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/services/codefixes/helpers.ts:366]: You had me at prettier config...

File: src/services/codefixes/helpers.ts [hunk 3]
```diff
@@ -349,7 +348,28 @@ namespace ts.codefix {
         }
     }
 
-    export function typeToAutoImportableTypeNode(checker: TypeChecker, importAdder: ImportAdder, type: Type, contextNode: Node | undefined, scriptTarget: ScriptTarget, flags?: NodeBuilderFlags, tracker?: SymbolTracker): TypeNode | undefined {
+    function createTypeParametersForArguments(argumentTypeParameters: string[], typeArguments: NodeArray<TypeNode> | undefined) {
+        const usedNames = new Set(argumentTypeParameters);
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/services/codefixes/helpers.ts]: It's so generated type parameter names don't have the same name as any existing type parameter. See `tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterArgumentDifferent.ts`'s fix output if this is switched to just an array:

```diff
- function added<T, U>(value: T, arg1: string) {
+ function added<T, T>(value: T, arg1: string) {
```

File: tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterConstrained.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+goTo.marker("1");
+verify.codeFix({
+  description: "Add missing function declaration 'added'",
+  index: 0,
+  newFileContent: `function existing<T extends string>(value: T) {
+  added(value);
+}
+
+function added<T>(value: T) {
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterConstrained.ts:16]: Hmm, yeah. And I just added another breaking test case with `incompleteFunctionCallCodefixTypeUnion.ts`:

```diff
- function added<T>(value: T) {
+ function added<T>(value: T | (U & string)) {
```

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterConstrained.ts:16]: This is trickier than I realized 😕. Will draft for a bit so I can try to think on how to get this to work. Thanks for the feedback!

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterConstrained.ts:16]: All right, now it stores type parameter constraints and adds them to the new created parameters.

File: tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterNarrowed.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/// <reference path='fourslash.ts' />
+
+// @noImplicitAny: true
+////function existing<T>(value: T) {
+////  if (typeof value === "number") {
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply tests/cases/fourslash/incompleteFunctionCallCodefixTypeParameterNarrowed.ts:5]: I love it. Added `incompleteFunctionCallCodefixTypeParameterVariable.ts`. Right now the parameter becomes `et: string` from the `checker.getBaseTypeOfLiteralType` widening in `typeToAutoImportableTypeNode`.

```ts
function added1(et: string) {
    throw new Error("Function not implemented.")
}
```

**Review by sandersn — APPROVED** [review]

**Pull request merged**